### PR TITLE
Added dynamic alias '@alias' in ActiveQuery

### PR DESCRIPTION
**Pull request by CedricYii**
| Q | A |
| --- | --- |
| Is bugfix? | no |
| New feature? | yes |
| Breaks BC? | no |
| Tests pass? | yes |
| Fixed issues | #7263, #10883 |

The `@alias` placeholder enables to address dynamicaly the alias of the current table name in an active query or in a relation (without knowing it in advance).

Example:

``` php
class Order {
...
  public static function find()
  {
    return parent::find()->orderBy( '@alias.id' );
  }

  //relation with table COMPANY
  public function getCustomer()
  {
    return $this->hasOne( Company::className(), ['id' => 'customerid'] )
                ->andOnCondition( [ '@alias.type' => 'customer' ] );
  }
...
}
```

Then

``` php
$orders = Order::find()->alias('o');
```

Build the query:

``` sql
select * from order o order by o.id
```

And

``` php
$orders = Order::find()->alias('o')->joinWith('customer c');
```

Build the query:

``` sql
select o.* from order o 
left join company c on c.id = o.customerid and c.type = 'customer'
order by o.id
```

### CODE REVIEW

**Comment by lynicidn** [comment]
return parent::find()->alias('ali')->orderBy( 'ali.id' );

**Comment by lynicidn** [comment]
where you want configure your aliases ?

File: framework/db/ActiveQuery.php [hunk 1]
```diff
@@ -81,6 +81,11 @@ class ActiveQuery extends Query implements ActiveQueryInterface
     const EVENT_INIT = 'init';
 
     /**
+     * string placeholder to be replaced by the table alias of a query.
+     */
+    const ALIAS_PLACEHOLDER = '@alias';
```
Comment (lynicidn) [inline framework/db/ActiveQuery.php]: what? once alias for all queries ?
try it:
A::find()->joinWth(['b' => function($q){$q->andWhere(['@alias.id' => 1])}])->andWhere(['@alias.id' => 2])

File: framework/db/ActiveQuery.php [hunk 2]
```diff
@@ -624,16 +635,16 @@ private function joinWithRelation($parent, $child, $joinType)
...
-            $this->andHaving($child->having);
+            $this->andHaving( $child->having );
         }
         if (!empty($child->orderBy)) {
-            $this->addOrderBy($child->orderBy);
+            $this->addOrderBy( $child->orderBy );
         }
         if (!empty($child->groupBy)) {
-            $this->addGroupBy($child->groupBy);
+            $this->addGroupBy( $child->groupBy );
```
Comment (lynicidn) [inline framework/db/ActiveQuery.php]: read code style concept

File: framework/db/ActiveQuery.php [hunk 3]
```diff
@@ -789,4 +800,69 @@ public function alias($alias)
         }
         return $this;
     }
+
+    /**
+     * Replace occurence of self::ALIAS_PLACEHOLDER by $aliasName into given conditions
+     * @param string|array $condition the given conditions
+     * @param string $aliasName
+     */
+    private function replaceAliasPlaceholder($condition, $aliasName)
```
Comment (lynicidn) [inline framework/db/ActiveQuery.php]: why is private? you work with private var else where profit of private method ?

File: framework/db/ActiveQuery.php [hunk 4]
```diff
@@ -789,4 +800,69 @@ public function alias($alias)
...
+     */
+    private function normalizeAliasInQuery()
+    {
+        $alias = $this->getAliasTableName();
+
+        $this->on       = self::replaceAliasPlaceholder( $this->on, $alias );
+        $this->where    = self::replaceAliasPlaceholder( $this->where, $alias );
+        $this->having   = self::replaceAliasPlaceholder( $this->having, $alias );
+        $this->orderBy  = self::replaceAliasPlaceholder( $this->orderBy, $alias );
+        $this->groupBy  = self::replaceAliasPlaceholder( $this->groupBy, $alias );
```
Comment (lynicidn) [inline framework/db/ActiveQuery.php]: select?

File: tests/data/ar/Item.php [hunk 5]
```diff
@@ -20,4 +20,11 @@ public function getCategory()
     {
         return $this->hasOne(Category::className(), ['id' => 'category_id']);
     }
+
+    //relation with dynamic alias
+    public function getCategoryWithAlias()
```
Comment (lynicidn) [inline tests/data/ar/Item.php]: has one relation exists, need write test for exists relations, then all test cover you suggestion

File: tests/data/ar/Order.php [hunk 6]
```diff
@@ -196,4 +218,14 @@ public function attributeLabels()
             'total' => 'Invoice Total',
         ];
     }
+
+
+    public static function findWithAlias()
```
Comment (lynicidn) [inline tests/data/ar/Order.php]: what what what? for what it?

File: tests/framework/db/ActiveRecordTest.php [hunk 7]
```diff
@@ -1250,4 +1250,77 @@ public function testLinkWhenRelationIsIndexed3()
...
+        $orders = $query->all();
+
+        $this->assertEquals(3, $orders[2]->id);
+        $this->assertEquals(3, count($orders));
+    }
+
+    /**
+     * Check if the dynamic alias @alias is replaced in a query a related joinWith
+     */
+    public function testAliasDynamicInJoin()
```
Comment (lynicidn) [inline tests/framework/db/ActiveRecordTest.php]: you think it case need test in active record? - this test for query

File: framework/db/ActiveQuery.php (hunk 1 above)
Comment (CedricYii) [inline framework/db/ActiveQuery.php]: Thanks for your feedback.
However, I have a hard time to understand what you are trying to point out. 
Could you be more explicit?

File: tests/framework/db/ActiveRecordTest.php (hunk 7 above)
Comment (CedricYii) [inline tests/framework/db/ActiveRecordTest.php]: How would you test it through Query (my commit is on ActiveQuery only)?

File: tests/data/ar/Item.php (hunk 5 above)
Comment (CedricYii) [inline tests/data/ar/Item.php]: I am testing a relation with a dynamic alias (that doesn't exist), not a "regular" relation.

File: framework/db/ActiveQuery.php (hunk 3 above)
Comment (CedricYii) [inline framework/db/ActiveQuery.php]: Ok, I'll change it

File: framework/db/ActiveQuery.php (hunk 4 above)
Comment (CedricYii) [inline framework/db/ActiveQuery.php]: Would be nice indeed. I had an unexpected behavior with select during development, I need to check if it's now ok with this last version.

**Comment by petrabarus** [comment]
This is really great! Awesome work @CedricYii!

**Comment by CedricYii** [comment]
@petrabarus Thank you for your comment! You are welcome.
Now that I've used this feature in a project, it sounds like it's a must have as it makes building complex ActiveQueries really easier.

**Comment by lynicidn** [comment]
what is real use case?
(new UserQuery())->select(['@alias.username, @alias.contact_phone'])->from('@alias')->with('profile') 
???

**Comment by petrabarus** [comment]
@lynicidn you can read one of the use case from me here #7263

File: framework/db/ActiveQuery.php [hunk 8]
```diff
@@ -806,7 +806,7 @@ public function alias($alias)
      * @param string|array $condition the given conditions
      * @param string $aliasName
      */
-    private function replaceAliasPlaceholder($condition, $aliasName)
+    public function replaceAliasPlaceholder($condition, $aliasName)
```
Comment (petrabarus) [inline framework/db/ActiveQuery.php]: why do you need this to be public?

Comment (CedricYii) [inline framework/db/ActiveQuery.php]: It's not needed but wouldn't hurt as @lynicidn suggested.
For some reasons I don't know, methods (removeDuplicatedModels, buildJoinWith, getQueryTableName, ...) are in private so I sticked to it for normalizeAliasInQuery. However I would have put it in protected.

**Comment by CedricYii** [comment]
@lynicidn It doesn't make sense to use this dynamic aliasing in a single query nor in directly a single activeQuery itself.
An simple example is provided with this PR.
Let's try to extend this example a little:
````php
class Order {
...
  public static function find()
  {
    return parent::find()->orderBy( '@alias.id' );
  }
...
}
````
So you can write,
````php
$orders = Order::find()->alias('o');
````
But you can also write in the same application using the same model
````php
$orders = Order::find()->alias('otherAlias');
````
Or even
````php
$orders = Order::find();
````
All three statements will give the same result (list of orders ordered by id) without any need to worry about what's in the model Order, nor to modify it nor to extend it not to overload it.
Does this make things clearer?

**Comment by lynicidn** [comment]
@CedricYii 
```php
class Order {
...
  public static function find()
  {
    return parent::find()->orderBy( '@alias.id' );
  }
...
}

class User {
...
  public static function find()
  {
    return parent::find()->orderBy( '@alias.id' );
  }
    public function getOrders() {
        $this->hasMany('Order', ['user_id' => 'id']);
    }
...
}
```
what is expected result where ? User::find()->with('orders') ?

**Comment by CedricYii** [comment]
@lynicidn: You will get
// SELECT * FROM user ORDER BY user.id
// SELECT * FROM order WHERE `user_id` IN (...) ORDER BY order.id

To change condition for relation, modify getOrders:
````php
class User {
...
    public function getOrders() {
        return $this
            ->hasMany('Order', ['user_id' => 'id'])
            ->andOnCondition('@alias.customer_id' => 15)
            ->orderBy(['@alias.id' => 'DESC']);
    }
...
}
````
Note that there are no conflict names in these queries, so `@alias` remains useless here.

**Comment by CedricYii** [comment]
@cebe Would it be possible to plan this in 2.0.9 as it provides a complete solution as we talked about?

**Comment by CedricYii** [comment]
Note that another practical usage is when you want to get extra columns for an ActiveRecord model.

For example, let's say you want to get the customer name with the order:
Before you needed to write:
````php
class Order {
...
  public static function find()
  {
    return parent::find()
            ->select([static::tableName().'*', 'c.name'])
            ->joinWith(['customer c'], false);
  }
...
}
````
Which was not really explicit and not portable (you couldn't define a new alias on this model).

Now you can just write:
````php
class Order {
...
  public static function find()
  {
    return parent::find()
            ->select(['@alias.*', 'c.name'])
            ->joinWith(['customer c'], false);
  }
...
}
````
Which looks more intuitive and straightforward, and you can reuse it with dynamic aliasing.

**Comment by CedricYii** [comment]
@cebe I have rebased the commits, could you give it a try to have it included in 2.0.10?

**Comment by CedricYii** [comment]
@cebe Let me know if I can help you to have this one merged?

**Comment by dynasource** [comment]
although I appreciate all the effort that is put into this solution, IMHO, it not yet ready for integration as:
* it bloats ActiveQuery and adding too much complexity to ActiveQuery
* it kills ActiveQuery performance

**Comment by CedricYii** [comment]
@dynasource Most of the time adding a feature means adding code, which one shouldn't be afraid of if the feature is adding value to the framework.
In this case, we are talking about a 9% inflation (72 lines out of 797 lines of code). And it does not change the way you are using ActiveQuery if you don't want to use it.
And if you end up using it one day, you will find that you couldn't go back coding without `@alias` as it greatly simplifies development. 
Or maybe I missed the point you are thinking about in your first sentence?

As for performance issue, have you done a benchmark or tried it and noticed consequent lags? 
Note that I am using str_replace only (not regex like for already merged table alias feature) to keep it fast.
However ActiveQuery is not the fastest way to build queries for sure, but it is the most flexible, maintainable and scalable; and `@alias` is going this way.

**Comment by CedricYii** [comment]
I did a benchmark on a page overloaded with 200 queries using dynamic `@alias` (and ActiveQuery with aliasDynamic) and the same one with hard-coded aliases (and original ActiveQuery), and it has been slowed down by only 3%.
Estimation trying to isolate the impact on ActiveQuery part itself is 15%.
So I would not say it kills performance in regard of the feature offered.

However it remains quite simple to desactivate this by overriding `ActiveQuery::normalizeAliasInQuery` or add an option to do so.

**Comment by dynasource** [comment]
> I did a benchmark on a page overloaded with 200 queries using dynamic @alias (and ActiveQuery with aliasDynamic) and the same one with hard-coded aliases (and original ActiveQuery), and it has been slowed down by only 3%.

thanks for your replies and doing a benchmark. 
Your benchmark unfortunately proves what I was afraid of:
* 3% performance loss on a new feature which will cover ~1% of usecases 

There it must be recoded in such way that if somebody want to use it, he/she can, but for default scenario's it must be left out (unless performance is in no way killed).

The reason why I care about performanceloss, even if its only 3% in your opinion, comes from 2 reasons: local testing + production. Both are extremely important for me to be optimal.

**Comment by antonmarin** [comment]
> will cover ~1% of usecases
@dynasource, why you think it's only about 1%? As i understand it as almoust every ActiveQuery method, that can be used to detail query by relative record.

**Comment by dynasource** [comment]
True. I misformulated that. It can indeed be used in every usecase, however, what I tried to point out, is that you will not always need this, but I does go at cost of 3% at all times.

**Comment by antonmarin** [comment]
I see now. @CedricYii do you have ideas how to implement placeholder without perfomance regression?

**Comment by Renkas** [comment]
Any progress on this? This feature is long overdue in framework :(

**Comment by samdark** [comment]
@Renkas no. Performance will be degraded if we'll merge it as is.

File: framework/db/ActiveQuery.php [hunk 9]
```diff
@@ -794,4 +805,64 @@ public function alias($alias)
...
+        return $condition;
+    }
+
+    /**
+     * Returns the table alias (or the table name if empty) of the current ActiveQuery
+     * @return string table alias or table name of the current activeQuery
+     */
+    public function getAliasTableName()
+    {
+        list (, $alias) = self::getQueryTableName($this);
```
Comment (Faryshta) [inline framework/db/ActiveQuery.php]: retun self::getQueryTableName($this)[1];

?

File: framework/db/ActiveQuery.php [hunk 10]
```diff
@@ -794,4 +805,64 @@ public function alias($alias)
...
+    public function replaceAliasPlaceholder($condition, $aliasName)
+    {
+        if ((empty($condition)) or (is_int($condition))) {
+            return $condition;
+        } elseif (is_string($condition)) {
+            return str_replace(self::ALIAS_PLACEHOLDER, $aliasName, $condition);
+        } elseif (is_array($condition)) {
+            $conditionNew = [];
+            foreach ($condition as $key => $value) {
+                unset($condition[$key]);
```
Comment (Faryshta) [inline framework/db/ActiveQuery.php]: redundant since line 828 rewrites `$condition` anyway

**Comment by Faryshta** [comment]
@samdark the method `normalizeAliasInQuery()` only gets called when building the actual query.

the travis build doesn' seem wildly affected either. although i can't see individual tests performance.

**Comment by Faryshta** [comment]
@CedricYii i was thinking to use another name for this alias such as `@main` to be able to use other automatic aliases later based on relation names. example

```php
User::find()->innerJoinWith('tickets')->andWhere([
     '@main.status' => User::STATUS_ACTIVE,
     '@tickets.status' => Ticket::STATUS_OPEN,
]);
```

all the inner joins would cause an automatic alias with the name of the relation being turned into the related table name and would allow nesting.

```php
Store::find()->innerJoinWith('sales.articles')->select([
    'brute_sales' => 'sum(@sales.articles.price)' 
]);
```

for this pr i think you just need to edit:

`const ALIAS_PLACEHOLDER = '@main';`

**Comment by Faryshta** [comment]
as for the performance, how about a method `ActiveQuery::dynamicAlias(boolean $active)` by default it will be unactivated and `normalizeAliasInQuery()` won't be executed. saving the query builder a whooping 3% performance.

**Comment by rob006** [comment]
Why we event care about this 3%? This is essential feature. Right now I don't even consider porting some apps from Yii 1.1 to Yii 2 because of poorly aliases support.

**Comment by Faryshta** [comment]
i am working on a pr which will be based on this but wont use `@alias` instead it will use `@main` for the main table and the relation name prefixed with `@` symbol will be tthe automatic alias for the a related record supporting nesting.

example

```php
class Order {
  public static function find()
  {
    return parent::find()->orderBy( '@main.id' );
  }

  //relation with table COMPANY
  public function getCustomer()
  {
    return $this->hasOne( Company::className(), ['id' => 'customerid'] )
                ->andOnCondition( [ '@customer.type' => 'customer' ] );
  }
}
```

nested example

```php
Store::find()->joinWith('sales.articles')
    ->where(['@main.id' => 1])
    ->sum('@sales.articles.amount')
```

**Comment by njasm** [comment]
any decisions on this one? /cc @CedricYii  @Faryshta @cebe @rob006 @samdark

**Comment by cebe** [comment]
this needs some serious amount of dedicated time to be handled correctly, which I currently do not have. Sorry. Its not going to make it into 2.0.13

**Comment by samdark** [comment]
https://travis-ci.org/yiisoft/yii2/jobs/338904472

**Comment by CedricYii** [comment]
@samdark I know but QueryBuilder in master seem to be broken anyway...  Not related to my commit.

**Comment by samdark** [comment]
It is not: https://travis-ci.org/yiisoft/yii2/builds/338703011

**Comment by CedricYii** [comment]
I see now. I thought 2.0.14 wasn't supposed to break BC, but it did in the QueryBuilder commit (d165863) 4 days ago... I will adapt accordingly.

**Comment by CedricYii** [comment]
@samdark Is it ok for you now?

**Comment by samdark** [comment]
@CedricYii I'm not into this issue so can't say about the whole concept w/o help from @cebe and @SilverFire.

**Comment by CedricYii** [comment]
@samdark Thank you. As the main concern about performance for those that may not use this feature is now addressed, would you mind letting it a chance in the 2.0 milestones (instead of the far far away 2.1.1 you set 9 days ago)?

File: framework/db/ActiveQuery.php (hunk 10 above)
Comment (CedricYii) [reply framework/db/ActiveQuery.php]: Thanks in removed it.

File: framework/db/ActiveQuery.php (hunk 9 above)
Comment (CedricYii) [reply framework/db/ActiveQuery.php]: Ok I changed it this way, but it's not the way it was done in the master.

**Comment by samdark** [comment]
@cebe said he'll check it soon.

**Comment by CedricYii** [comment]
@Faryshta Concerning your proposal about using relation name as the alias placeholder, I am not sure to see the advantages. As far as I understand, it would not work when the same relation is used twice, it would be harder to maintain (if we change the name of the relation, the placeholder must be changed too) and it would be slower. I feel like it means using an alias of the alias of a relation.
Your nested example could be handled by alias in nested relation #11421.

**Comment by samdark** [comment]
@CedricYii discussed within code team and decided to move it to 2.1.

**Comment by CedricYii** [comment]
@samdark Too bad, thanks for the feedback anyway.

**Comment by MattiCerio** [comment]
Thanks for the work, this would be useful!
At the moment there is a temporary solution to retrieve in the relation definition the alias used ?

**Comment by samdark** [comment]
Won't be merged into 2.x.

**Pull request closed without merging**